tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php

```diff
@@ -0,0 +1,121 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Tests\Metadata\Property\Factory;
+
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Property\Factory\SerializerPropertyMetadataFactory;
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Tests\Fixtures\TestBundle\Entity\Dummy;
+use ApiPlatform\Core\Tests\Fixtures\TestBundle\Entity\RelatedDummy;
+use Symfony\Component\PropertyInfo\Type;
+use Symfony\Component\Serializer\Mapping\AttributeMetadata as SerializerAttributeMetadata;
+use Symfony\Component\Serializer\Mapping\ClassMetadata as SerializerClassMetadata;
+use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactoryInterface as SerializerClassMetadataFactoryInterface;
+
+/**
+ * @author Teoh Han Hui <teohhanhui@gmail.com>
+ */
+class SerializerPropertyMetadataFactoryTest extends \PHPUnit_Framework_TestCase
+{
+    public function testConstruct()
+    {
+        $resourceMetadataFactoryProphecy = $this->prophesize(ResourceMetadataFactoryInterface::class);
+        $resourceMetadataFactory = $resourceMetadataFactoryProphecy->reveal();
+
+        $serializerClassMetadataFactoryProphecy = $this->prophesize(SerializerClassMetadataFactoryInterface::class);
+        $serializerClassMetadataFactory = $serializerClassMetadataFactoryProphecy->reveal();
+
+        $decoratedProphecy = $this->prophesize(PropertyMetadataFactoryInterface::class);
+        $decorated = $decoratedProphecy->reveal();
+
+        $serializerPropertyMetadataFactory = new SerializerPropertyMetadataFactory($resourceMetadataFactory, $serializerClassMetadataFactory, $decorated);
+
+        $this->assertInstanceOf(PropertyMetadataFactoryInterface::class, $serializerPropertyMetadataFactory);
+    }
+
+    public function testCreate()
+    {
+        $resourceMetadataFactoryProphecy = $this->prophesize(ResourceMetadataFactoryInterface::class);
+        $dummyResourceMetadata = (new ResourceMetadata())
+            ->withAttributes([
+                'normalization_context' => [
+                    'groups' => ['dummy_read'],
+                ],
+                'denormalization_context' => [
+                    'groups' => ['dummy_write'],
+                ],
+            ]);
+        $resourceMetadataFactoryProphecy->create(Dummy::class)->willReturn($dummyResourceMetadata)->shouldBeCalled();
+        $resourceMetadataFactoryProphecy->create(RelatedDummy::class)->willReturn(new ResourceMetadata())->shouldBeCalled();
+        $resourceMetadataFactory = $resourceMetadataFactoryProphecy->reveal();
+
+        $serializerClassMetadataFactoryProphecy = $this->prophesize(SerializerClassMetadataFactoryInterface::class);
```

Comment:
This whole unreadable chunk here is due to the lack of fluent interface in Symfony Serializer :smile: